regidx.c

```diff
@@ -244,48 +329,136 @@ void regidx_destroy(regidx_t *idx)
     free(idx);
 }
 
-int regidx_overlap(regidx_t *idx, const char *chr, hts_pos_t from, hts_pos_t to, regitr_t *itr)
+static int reglist_build_index_(regidx_t *regidx, reglist_t *list)
 {
-    if ( itr ) itr->i = itr->n = 0;
+    int i;
+    if ( list->unsorted ) {
+        if ( !regidx->payload_size ) {
+            qsort(list->reg,list->nreg,sizeof(reg_t),cmp_reg_ptrs);
+        } else {
+            reg_t **ptr = malloc(sizeof(*ptr)*list->nreg);
+            if (!ptr) return -1;
+            for (i=0; i<list->nreg; i++) ptr[i] = list->reg + i;
+            qsort(ptr,list->nreg,sizeof(*ptr),cmp_reg_ptrs2);
+
+            uint8_t *tmp_dat = malloc(regidx->payload_size*list->nreg);
+            if (!tmp_dat) { free(ptr); return -1; }
+            for (i=0; i<list->nreg; i++) {
+                size_t iori = ptr[i] - list->reg;
+                memcpy(tmp_dat+i*regidx->payload_size,
+                       list->dat+iori*regidx->payload_size,
+                       regidx->payload_size);
+            }
+            free(list->dat);
+            list->dat = tmp_dat;
+
+            reg_t *tmp_reg = (reg_t*) malloc(sizeof(reg_t)*list->nreg);
+            if (!tmp_reg) { free(ptr); return -1; }
+            for (i=0; i<list->nreg; i++) {
+                size_t iori = ptr[i] - list->reg;
+                tmp_reg[i] = list->reg[iori];
+            }
+            free(ptr);
+            free(list->reg);
+            list->reg  = tmp_reg;
+            list->mreg = list->nreg;
+        }
+        list->unsorted = 0;
+    }
 
-    int iseq;
-    if ( khash_str2int_get(idx->seq2regs, chr, &iseq)!=0 ) return 0; // no such sequence
+    list->nidx = 0;
+    int j,k, midx = 0;
+    for (j=0; j<list->nreg; j++) {
+        int ibeg = iBIN(list->reg[j].beg);
+        int iend = iBIN(list->reg[j].end);
+        if ( midx <= iend )
+        {
+            int old_midx = midx;
+            midx = iend + 1;
+            kroundup32(midx);
+            uint32_t *new_idx = realloc(list->idx, midx*sizeof(uint32_t));
+            if (!new_idx) return -1;
+            list->idx = new_idx;
+            memset(list->idx+old_midx, 0, sizeof(uint32_t)*(midx-old_midx));
+        }
+        if ( ibeg==iend ) {
+            if ( !list->idx[ibeg] ) list->idx[ibeg] = j + 1;
+        } else {
+            for (k=ibeg; k<=iend; k++)
+                if ( !list->idx[k] ) list->idx[k] = j + 1;
+        }
+        if ( list->nidx < iend+1 ) list->nidx = iend+1;
+    }
```

Comment:
This does not seem safe for 64-bit positions.

Reply:
Agreed. reg_t is defined as:

```
#define iBIN(x) ((x)>>13)

typedef struct
{
    hts_pos_t beg, end;
}
reg_t;
```

Hence ibeg and iend should be declared hts_pos_t above too.  Good spot.